File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculator/wheel/GravityAccelerationCalculator.java
package com.github.aadvorak.artilleryonline.battle.calculator.wheel;

import com.github.aadvorak.artilleryonline.battle.calculations.WheelCalculations;
import com.github.aadvorak.artilleryonline.battle.calculations.WheelGroundState;

public class GravityAccelerationCalculator {

    public static void calculate(WheelCalculations wheelCalculations, double roomGravityAcceleration,
                                 double groundGravityDepth) {
        if (WheelGroundState.FULL_OVER_GROUND.equals(wheelCalculations.getGroundState())) {
            wheelCalculations.getGravityAcceleration()
                    .setX(0.0)
                    .setY(-roomGravityAcceleration);
        } else if (WheelGroundState.HALF_OVER_GROUND.equals(wheelCalculations.getGroundState())
                && wheelCalculations.getGroundDepth() <= groundGravityDepth) {
            var groundAngle = wheelCalculations.getGroundAngle();
            var groundAccelerationModule = Math.abs(roomGravityAcceleration * Math.sin(groundAngle))
                    * Math.sqrt(1 - wheelCalculations.getGroundDepth() / groundGravityDepth);
            wheelCalculations.getGravityAcceleration()
                    .setX(-groundAccelerationModule * Math.sin(groundAngle))
                    .setY(-groundAccelerationModule * Math.cos(groundAngle));
        }
    }
}

File: frontend/playground/battle/calculator/wheel/gravity-acceleration-calculator.ts
import { type WheelCalculations, WheelGroundState } from '@/playground/data/calculations'

export const GravityAccelerationCalculator = {
  calculate(
      wheelCalculations: WheelCalculations,
      roomGravityAcceleration: number,
      groundGravityDepth: number
  ): void {
    if (WheelGroundState.FULL_OVER_GROUND === wheelCalculations.groundState) {
      wheelCalculations.gravityAcceleration.x = 0.0
      wheelCalculations.gravityAcceleration.y = -roomGravityAcceleration
    } else if (
      WheelGroundState.HALF_OVER_GROUND === wheelCalculations.groundState &&
      wheelCalculations.groundDepth! <= groundGravityDepth
    ) {
      const groundAngle = wheelCalculations.groundAngle!
      const groundAccelerationModule = Math.abs(roomGravityAcceleration * Math.sin(groundAngle))
      wheelCalculations.gravityAcceleration.x = -groundAccelerationModule * Math.sin(groundAngle)
      wheelCalculations.gravityAcceleration.y = -groundAccelerationModule * Math.cos(groundAngle)
    }
  }
}
